Add CMAKELISTS recommendations (#27)
beman-project/beman: docs/beman-standard.md

Its goal is to create consistency facilitating the evaluation of, and
contribution to Beman libraries.

<!-- toc -->

## Table of contents

- [The Beman Standard](#the-beman-standard)
  - [Table of contents](#table-of-contents)
  - [Introduction](#introduction)
    - [Core principles](#core-principles)
    - [Changing this document](#changing-this-document)
    - [Conventions](#conventions)
  - [License](#license)
  - [General](#general)
  - [Top-level](#top-level)
  - [`README.md`](#readmemd)
  - [`CMakeLists.txt`](#cmakeliststxt)
  - [Directory Layout](#directory-layout)
  - [File contents](#file-contents)

> <!-- tocstop -->

Author: Yes. Why?

Also run markdownlint over the file to be sure it's OK. It seems similar to what we have in `Beman/Beman.Beman/docs/CODE_OF_CONDUCT.md` (including comment)

> <!-- tocstop -->

Reviewer: How is it generated? If we do a table of contents, I'd like to understand the tooling required. Before this PR, it required only a text editor.

> <!-- tocstop -->

Author: <img width="762" alt="image" src="https://github.com/user-attachments/assets/32ef07c2-26da-4e37-811f-465fc0c0e217">
I use this plugin for VSCode which seems to generate the exact format from ToC in `Beman/Beman.Beman/docs/CODE_OF_CONDUCT.md`.

> <!-- tocstop -->

Reviewer: The editor I use (spacemacs) can also generate a table of contents, but it does so with different syntax. I don't think we should force everyone to use the same editor so I'd rather we not include a table of contents.

> <!-- tocstop -->

Reviewer: Additionally, GitHub already provides an outline folks can use to navigate the document:

![image](https://github.com/user-attachments/assets/a60c4ca1-20e3-4f68-b573-661772c243ad)

> <!-- tocstop -->

Reviewer 2: I vote for keeping it simple here and circling back. Note that we could convert to reStructuredText, which support a [toctree annotation](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-toctree) for expanding TOC. No fancy generator tech needed. Try rst out in a gist to see if you like it.

> <!-- tocstop -->

Author: @bretbrownjr , @camio , I agree we can keep a simple solution. Previously, I've showed that what's in this PR is existing practice in `Beman/Beman.Beman/docs/CODE_OF_CONDUCT.md`. As per my understanding in this thread, I removed autogenerated ToCs in both files.

## `CMakeLists.txt`

**[CMAKE.DEFAULT]** RECOMMENDATION: The root `CMakeLists.txt` should build all targets by default (including dependency targets).

**[CMAKE.SKIP_TESTS]** RECOMMENDATION: The root `CMakeLists.txt` should not build tests and their dependencies when `BUILD_TESTING` is set to `OFF` (see [CTest docs](https://cmake.org/cmake/help/latest/module/CTest.html)). Use the following style:

```CMake
# <repo>/CMakeLists.txt
# ...
if(BUILD_TESTING)
  FetchContent_Declare(
    googletest
    GIT_REPOSITORY https://github.com/google/googletest.git
    GIT_TAG f8d7d77c06936315286eb55f8de22cd23c188571 # release-1.14.0
  )
  FetchContent_MakeAvailable(googletest)
endif()

# ...
if(BUILD_TESTING)

>   add_subdirectory(/path/to/tests) # CMAKE.AVOID_PASSTHROUGHS already applied.

Reviewer: ```suggestion
  add_subdirectory(/path/to/tests)
```

Suggest removing the comment since it doesn't apply to this entry and may confuse folks.

contain a one- or two-paragraph summary describing the library's purpose.

**[README.IMPLEMENTS]** RECOMMENDATION: Following the purpose and a newline, the
`README.md` should indicate which papers the repository implements. Use the following style:

```markdown
**Implements:** [`std::optional<T&>` (P2988R5)](https://wg21.link/P2988R5) and
[Give *std::optional* Range Support (P3168R1)](https://wg21.link/P3168R1).
```

## `CMakeLists.txt`

> **[CMAKELISTS.DEFAULT]** RECOMMENDATION: The root `CMakeLists.txt` should build all targets by default (including all targets of its dependencies).

Reviewer: ```suggestion
**[CMAKELISTS.DEFAULT]** RECOMMENDATION: The root `CMakeLists.txt` should build all targets by default (including dependency targets).
```

> **[CMAKELISTS.SKIP_TESTS]** RECOMMENDATION: The root `CMakeLists.txt` should not build tests and their dependencies when `BUILD_TESTING` is set to `OFF` (check [CTest docs](https://cmake.org/cmake/help/latest/module/CTest.html)). Use the following style:

Reviewer: ```suggestion
**[CMAKELISTS.SKIP_TESTS]** RECOMMENDATION: The root `CMakeLists.txt` should not build tests and their dependencies when `BUILD_TESTING` is set to `OFF` (see [CTest docs](https://cmake.org/cmake/help/latest/module/CTest.html)). Use the following style:
```

[Give *std::optional* Range Support (P3168R1)](https://wg21.link/P3168R1).
```

## `CMakeLists.txt`

> **[CMAKELISTS.DEFAULT]** RECOMMENDATION: The root `CMakeLists.txt` should build all targets by default.

Author: Check again.

**[CMAKELISTS.SKIP_TESTS]** RECOMMENDATION: The root `CMakeLists.txt` should allow disabling tests build via CMake argument `BUILD_TESTING=OFF` (check [CTest docs](https://cmake.org/cmake/help/latest/module/CTest.html)). Follow the

> following example's style:

Author: I am OK with your suggestion but I literally copied `Follow the following example's style` from a previous section. Should I apply your suggestion everywhere?

> following example's style:

Reviewer: LOL, if you don't mind.

> following example's style:

Author: fixed

- `CMakeLists.txt` snippets:

> ```Makefile

Reviewer: ```suggestion
```CMake
```
